pyccel/ast/basic.py

```diff
@@ -23,13 +27,217 @@ def __new__(cls, *args, **kwargs):
         hashable_args  = [a if not isinstance(a, list) else tuple(a) for a in args]
         return sp_Basic.__new__(cls, *hashable_args)
 
+    def __init__(self):
+        self._user_nodes = []
+        self._fst = []
+        for c_name in self._attribute_nodes:
+            c = getattr(self, c_name)
+            from pyccel.ast.internals import PyccelSymbol
+            from pyccel.ast.literals import convert_to_literal
+            if isinstance(c, PyccelSymbol):
+                # Anti-pattern
+                # PyccelSymbol is not a Basic so it must be handled separately
+                continue
+            elif isinstance(c, (int, float, complex, str, bool)):
+                # Convert basic types to literal types
+                c = convert_to_literal(c)
+                setattr(self, c_name, c)
+
+            elif isinstance(c, iterable_types):
+                if any(isinstance(ci, iterable_types) for ci in c):
+                    raise TypeError("Basic child cannot be a tuple of tuples")
+                c = tuple(ci if (not isinstance(ci, (int, float, complex, str, bool)) \
+                                 or isinstance(ci, PyccelSymbol)) \
+                        else convert_to_literal(ci) for ci in c)
+                setattr(self, c_name, c)
+
+            elif not isinstance(c, Basic) and c is not None:
+                raise TypeError("Basic child must be a Basic or a tuple not {}".format(type(c)))
+
+            if isinstance(c, tuple):
+                for ci in c:
+                    if ci:
+                        ci.user_nodes = self
+            elif c:
+                c.user_nodes = self
+
+    def get_user_nodes(self, search_type, excluded_nodes = ()):
+        """ Find out if any of the user nodes are instances
+        of the provided object.
+
+        Parameters
+        ----------
+        search_type : ClassType or tuple of ClassTypes
+                      The types which we are looking for
+        excluded_nodes : tuple of types
+                      Types for which get_user_nodes should not be called
+
+        Results
+        -------
+        Boolean : True if one of the user nodes is an instance of
+                  the class in the argument
```

Comment:
your are returning list of nodes , maybe you need to update doc

Reply:
Fixed in 9a1a43ee